### ISSUE TITLE
Skip the merge queue when possible

### ISSUE THREAD

**Issue by zmitchell** [flox/flox#1676]
The purpose of the merge queue is to let the merge button be "fire and forget" e.g. let multiple jobs try to merge at the same time, and ensure that each one still passes the test suite after a job ahead in the queue has merged.

When there are no jobs in the queue and the PR is up to date, there's no reason to re-run the test suite, but GitHub Actions will still run the test suite again anyway. There are ways to get around this that we can employ to skip the merge queue entirely and save both developer time and billable time for our runners.

See this comment for some suggestions: https://github.com/orgs/community/discussions/43988#discussioncomment-9667914

In short:
- Check whether the PR branch is up to date with `main`
- Check how many jobs are in the merge queue using the GitHub API

**Comment by ysndr** [comment]
I think the comment thread above would not work in our situation.
We have configured required checks and skipping a job does not count as a successful check.
Thus, blank succeeding on pr (as in the comment) or the merge queue is not possible, unless you add a conditional _to every single step_ and add a "success step" at the end of every job which seems pretty tedious.

@garbas do you have another idea?

**Comment by garbas** [comment]
I would propose to skip at least the "*-dev" jobs when running in merge queue.

If we skip the job via `if: github.event_name != 'merge_group'` even if required, it is going to be skipped when calculating final check status of the workflow 
(see https://github.com/orgs/community/discussions/46757?sort=old#discussioncomment-4909336)

**Comment by garbas** [comment]
Generally, I would just not run anything (or maybe just nix-build to populate the binary cache) on merge queue.

**Comment by zmitchell** [comment]
I'm not sure I care how tedious it is to add a success condition to every step. It's YAML, you don't have a programming language at your disposal, and you're working inside someone else's framework (GitHub's), and sometimes that means doing something dirty or hacky.

Just compute the success condition up front (if merge queue is empty, we have an approval, and the branch is up to date), and skip all possible jobs given the success condition.

**Comment by dcarley** [comment]
I also don't feel great about the added complexity. Especially if we have to call out to GitHub's API, use GraphQL, and consider about race conditions.

Some other options that might achieve the same goal of reducing approval->shipping and billable hours:

1. Replace the workflow run on `main` by running `trigger-flox-installers-workflow` from the merge queue if all other jobs were successful. Functionally it should be the same, as we have the same commit ref in both runs, but it would be harder to navigate to the completed workflow for a commit on `main`.
1. Replace the merge queue with requiring branches to be up-to-date before merge. However it would cause more manual intervention if there are several consecutive merges and I don't recall whether the update button in the GitHub UI also triggers dismiss stale reviews (which I'd like to talk about sometime separately) which would make it even worse.

**Comment by zmitchell** [comment]
I would rather just get rid of the merge queue in that case

**Comment by zmitchell** [comment]
Putting this in the blocked column until we reach consensus on what behavior we actually want. Right now these are the concerns:
- Disabling the merge queue is only _safe_ if we require branches be up to date
- Keeping branches up to date with current settings will automatically dismiss an approval when a PR is rebased or `main` is merged into it.
- Tracking down an approval after an update is a burden.
- Allowing "stale" reviews to stand allows code to merge that hasn't been completely reviewed.

**Comment by mkenigs** [comment]
Discussed and won't disable for now
